### PR TITLE
Add fuzzing support

### LINKED ISSUE #115: Add fuzz testing
Fuzzing raw input is generally easier, but fuzzing the decisions would be more useful (but probably more involved/intrusive).

Some potentially helpful links:

https://github.com/pingcap/fail-rs
https://rust-fuzz.github.io/book/introduction.html
https://github.com/BurntSushi/quickcheck
https://github.com/altsysrq/proptest

and might as well add this link from @lu-zero for posterity:
https://doc.rust-lang.org/std/num/struct.Wrapping.html

### PR DESCRIPTION
**Pull request by YaLTeR**
Find assertion fails in your own rav1e clone in one simple command.

Closes #115, although new issues could be opened for adding, for example, proptest, or the randomizing RDO decisions thing.

The design I chose here (putting actual fuzzable functions inside the `rav1e` crate) should make it easy to hook up other fuzzers such as AFL and honggfuzz (more fuzzers more better), although it doesn't make it easy to hook up proptest or quickcheck because those frameworks want you to generate each individual parameter through them, rather than feeding in a chunk of random bytes (this is so that they can meaningfully shrink and expand the inputs, which is the whole point).

Additionally included is the ability to inspect crashing state straight through the fuzz target executables via the `RUST_LOG=debug` environment variable, which is extremely handy for figuring out what's up and making test cases.

Thus far trophies include everything fixed in #1619 and #1630 and a more-interesting case for #1611.

### CODE REVIEW

File: src/fuzzing.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  config.enc.bitrate = g.g();
+  config.enc.tile_cols = g.g();
+  config.enc.tile_rows = g.g();
+  config.enc.tiles = g.g();
+  config.enc.rdo_lookahead_frames = g.g();
+  config.enc.speed_settings = SpeedSettings::from_preset(g.g());
+  config.enc.show_psnr = g.g();
+  config.enc.train_rdo = g.g();
+
+  if debug {
```
Comment (tmatth) [inline src/fuzzing.rs]: Instea of having these `if debug` blocks, would it possible to wrap this in a macro or logging function that is a noop if the debug flag is disabled?

Comment (shssoichiro) [reply src/fuzzing.rs]: ^ Now that we have the `log` crate, you can use the `debug!` macro instead of `println!`.

Comment (shssoichiro) [reply src/fuzzing.rs]: ^ Now that we have the log crate in rav1e, you can use the `debug!` macro in place of `println!` to do this.

**Comment by YaLTeR** [comment]
Changed to use `debug!()`.

File: src/test_encode_decode/mod.rs [hunk 2]
```diff
@@ -38,7 +41,7 @@ fn fill_frame<T: Pixel>(ra: &mut ChaChaRng, frame: &mut Frame<T>) {
   }
 }
 
-pub(crate) fn read_frame_batch<T: Pixel>(
+fn read_frame_batch<T: Pixel>(
```
Comment (tmatth) [inline src/test_encode_decode/mod.rs]: Just out of curiosity, why did the visibility of this function change? Not seeing it called anywhere.

Comment (YaLTeR) [reply src/test_encode_decode/mod.rs]: Well, it is perhaps out of scope of this PR, but I noticed that it didn't need `pub(crate)` so I switched it to `pub`.

Comment (tmatth) [reply src/test_encode_decode/mod.rs]: I think the preference is for `pub(crate)` as we want to expose less API outside of the crate.

Comment (YaLTeR) [reply src/test_encode_decode/mod.rs]: Wait. I switched it to just `fn`, it's less visibility than `pub(crate)`.

Comment (tmatth) [reply src/test_encode_decode/mod.rs]: > Wait. I switched it to just `fn`, it's less visibility than `pub(crate)`.

Aha, I should have read your code and not your comment.

**Pull request merged**